// @google/dscc-scripts/build/viz/message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @license
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const bluebird = require("bluebird");
const CopyWebpackPlugin = require("copy-webpack-plugin");
const fs = require("mz/fs");
const path = require("path");
const webpack = require("webpack");
const args_1 = require("../args");
const util_1 = require("../util");
const util = require("./util");
const buildOptions = (buildValues, args) => {
    let transformString;
    const format = args.format;
    switch (format) {
        case args_1.MessageFormat.OBJECT:
            transformString = 'objectTransform';
            break;
        case args_1.MessageFormat.TABLE:
            transformString = 'tableTransform';
            break;
        default:
            return util_1.assertNever(format);
    }
    return {
        mode: 'development',
        entry: {
            // this is the viz source code
            main: path.resolve(__dirname, '../../', 'viz', 'printMessage.js'),
        },
        output: {
            filename: 'index.js',
            path: path.resolve(buildValues.pwd, 'build'),
        },
        plugins: [
            new CopyWebpackPlugin([
                {
                    from: path.resolve(buildValues.pwd, 'src', buildValues.jsonFile),
                    to: '.',
                },
                { from: path.join('src', buildValues.cssFile), to: '.' },
            ]),
            new webpack.DefinePlugin({
                TRANSFORM_PARAM: `"${transformString}"`,
            }),
        ],
    };
};
exports.buildMessage = async (args) => {
    const buildValues = util.validateBuildValues(args);
    const encoding = 'utf-8';
    const webpackOptions = buildOptions(buildValues, args);
    const compiler = webpack(webpackOptions);
    const compilerRun = bluebird.promisify(compiler.run, { context: compiler });
    await compilerRun();
    // put everything together except the manifest
    // once build directory is created...
    const manifestSrc = path.resolve(process.env.PWD, 'src', buildValues.manifestFile);
    const manifestDest = path.resolve(process.env.PWD, 'build', buildValues.manifestFile);
    const manifestContents = await fs.readFile(manifestSrc, encoding);
    const newManifest = manifestContents
        .replace(/YOUR_GCS_BUCKET/g, buildValues.gcsBucket)
        .replace(/"DEVMODE_BOOL"/, `${buildValues.devMode}`);
    return fs.writeFile(manifestDest, newManifest);
};
